tutorials/torchao_vllm_integration.md

```diff
@@ -0,0 +1,465 @@
+    group_size=128,
+    use_hqq=True,
+)
+assert isinstance(config, AOBaseConfig)
+```
+
+### 3. Module-Level Configuration
+
+For granular control, use `ModuleFqnToConfig`:
+
+```python
+from torchao.quantization import ModuleFqnToConfig, Int4WeightOnlyConfig, Int8WeightOnlyConfig
+
+config = ModuleFqnToConfig({
+    "model.layers.0.self_attn.q_proj": Int4WeightOnlyConfig(group_size=64),
+    "model.layers.0.self_attn.k_proj": Int4WeightOnlyConfig(group_size=64),
+    "model.layers.0.mlp.gate_proj": Int8WeightOnlyConfig(),
+    "_default": Int4WeightOnlyConfig(group_size=128)  # Default for other modules
+})
+```
+
+## Usage Examples
+
+### 1. Quantizing Models with HuggingFace Integration
+
+```python
+from transformers import TorchAoConfig, AutoModelForCausalLM
+from torchao.quantization import Int4WeightOnlyConfig
+
+# Create quantization configuration
+quantization_config = TorchAoConfig(
+    quant_type=Int4WeightOnlyConfig(group_size=128, use_hqq=True)
+)
+
+# Load and automatically quantize the model
+model = AutoModelForCausalLM.from_pretrained(
+    "meta-llama/Llama-3.2-1B",
+    torch_dtype="auto",
+    device_map="auto",
+    quantization_config=quantization_config
+)
+
+# Save quantized model
+model.push_to_hub("your-username/Llama-3.2-1B-int4", safe_serialization=False)
+```
+
+### 2. Serving with VLLM
+
+```bash
+# Start VLLM server with TorchAO quantized model
+vllm serve your-username/Llama-3.2-1B-int4 \
+    --quantization torchao \
+    --dtype bfloat16 \
+```
+
+
+## Adding New Quantization Methods to VLLM
+
+### Minimal Requirements for VLLM Compatibility
+
+To make a new TorchAO quantization method work with VLLM, you need to implement minimal tensor subclass operations that support **tensor parallelism**. VLLM uses `narrow()` and copy_ to move data from host cpu loaded in a state dict to the device, these  require these specific aten operations:
+
+```python
+@implements([aten.detach.default, aten.alias.default])
+def _(func, types, args, kwargs):
+    return return_and_correct_aliasing(
+        func, args, kwargs, args[0]._apply_fn_to_data(func)
+    )
+
+@implements(aten.copy_.default)
+def _(func, types, args, kwargs):
+    self = args[0]
+    src = args[1]
+    if _same_metadata(self, src):
+        self_tensors = self.__tensor_flatten__()[0]
+        for tensor_name in self_tensors:
+            getattr(self, tensor_name).copy_(getattr(src, tensor_name))
+        return
+    raise ValueError(
+        f"Not supported args for copy_ due to metadata mismatch: {args[0], args[1]}"
+    )
+
+@implements([aten.slice.Tensor])
+def _(func, types, args, kwargs):
+    "Example impl for weird quant scheme which has 1 scale per data element"
+    self, dim, start, end, step = fill_defaults(args, 5, [0, None, None, 1])
+    assert step == 1, f"Only step=1 is supported, got step={step}"
+    if end >= self.shape[dim]:
+        end = self.shape[dim]
+    shape = list(self.shape)
+    shape[dim] = end - start
```

Comment:
There's a lot of code in this tutorial. I feel the level of detail here is a bit distracting from the overall flow. What do you think about having a separate .py file with an actual runnable example, and then referencing that in the tutorial with some omitted code?